hazelcast/src/main/java/com/hazelcast/spi/impl/AbstractInvocationFuture.java

```diff
@@ -201,90 +658,450 @@ public final V get(final long timeout, final TimeUnit unit)
+                                                    @Nonnull Executor executor,
+                                                    @Nonnull CompletableFuture<Void> future) {
+        final Object value = resolve(state);
+        if (cascadeException(value, future)) {
+            return future;
+        }
         try {
-            executor.execute(new Runnable() {
-                @Override
-                public void run() {
-                    try {
-                        Object value = resolve(state);
-                        if (value instanceof Throwable) {
-                            Throwable error = unwrap((Throwable) value);
-                            callback.onFailure(error);
-                        } else {
-                            callback.onResponse((V) value);
-                        }
-                    } catch (Throwable cause) {
-                        logger.severe("Failed asynchronous execution of execution callback: " + callback
-                                + " for call " + invocationToString() + ": " + cause, cause);
-                    }
+            executor.execute(() -> {
+                try {
+                    consumer.accept((V) value);
+                    future.complete(null);
+                } catch (Throwable t) {
+                    future.completeExceptionally(t);
                 }
-
             });
         } catch (RejectedExecutionException e) {
-            callback.onFailure(wrapToInstanceNotActiveException(e));
+            future.completeExceptionally(wrapToInstanceNotActiveException(e));
+        }
+        return future;
+    }
+
+    /**
+     *
+     * @param waiter    the current wait node, see javadoc of {@link #state state field}
+     * @param executor  the {@link Executor} on which to execute the action associated with {@code waiter}
+     */
+    @SuppressWarnings("checkstyle:CyclomaticComplexity")
+    protected void unblockOtherNode(Object waiter, Executor executor) {
+        if (!(waiter instanceof Waiter)) {
+            return;
+        }
+        Object value = resolve(state);
+        if (waiter instanceof UniWaiter) {
+            ((UniWaiter) waiter).execute(executor, value);
+        } else if (waiter instanceof BiWaiter) {
+            Throwable t = (value instanceof ExceptionalResult) ? ((ExceptionalResult) value).cause : null;
+            value = (value instanceof ExceptionalResult) ? null : value;
+            ((BiWaiter) waiter).execute(executor, value, t);
+        } else if (waiter instanceof ExceptionallyNode) {
+            ((ExceptionallyNode) waiter).execute(value);
         }
     }
 
     protected abstract Exception wrapToInstanceNotActiveException(RejectedExecutionException e);
 
     // this method should not be needed; but there is a difference between client and server how it handles async throwables
-    protected Throwable unwrap(Throwable throwable) {
+    protected Throwable unwrap(ExceptionalResult result) {
```

Comment:
This was the method that causing the behavioral difference between client and server that is explained here 
https://github.com/hazelcast/hazelcast/issues/10335
I could not find any usages of it. IS it leftover ? should it be deleted along with the method overrides it in ClientInvocationFuture

Reply:
Yes, it's a leftover. Pushed a commit that removes `unwrap` method.